[3, 4단계 - 체스] 몰리(김지민) 미션 제출합니다.

jminkkk opened the pull request:
안녕하세요 현구막!

3,4단계 코드리뷰도 잘 부탁드립니다..! ~~(3, 4단계 진행 분리를 개인적으로만 진행했네요… 말씀을 잘못 이해했어요 🥲)~~

3단계 때 먼저 변경을 하고 4단계를 하다보니 생각보다 변경이 없긴 했지만 뭔가 계속 아쉬워서 수정을 하다보니 제출이 늦어진 것 같아요

저번 리뷰 때 객체지향 관련해서 조언을 부탁드렸었는데요..!

스스로 정답이 계속 없다고 생각해서 더 막막하게 느껴졌었는데, 현구막의 정답은 있다고 생각하신다는 말이 힘이 됐습니다 ㅎㅎ

## 설계 바뀐 부분

저번 코드리뷰를 통해 유지보수에 대해 고민을 해보았는데요! 

유지보수를 고려할 때, 항상 저의 관점에서의 유지보수를 생각하고 재사용성을 높이는 것에만 초점을 맞추었던 것 같습니다.

그래서 이번 리뷰에는 어느정도 재사용성을 포기해보자 라는 생각을 했습니다.

### 각 Piece에 대한 객체

> 저번 리뷰 때 `StraightCaptureObstacleRule` 보다는 오히려 `RookObstacleRule` 같은 방식이 오히려 룩에 대한 `ObstacleRule` 를 쉽게 인지할 수 있을 것 같다고 말씀해주셨는데요..! [관련 리뷰 링크](https://github.com/woowacourse/java-chess/pull/661#discussion_r1542193102)
<img width="682" alt="image" src="https://github.com/woowacourse/java-chess/assets/102847513/1afa211d-cfea-4fe4-9ef3-31aff90b335f">

+ `Piece` (같은 역할을 기대)를 상속, `Attack`(특정 동작을 기대)라는 인터페이스를 구현한 `DirectionAttackPiece` , `DiagonalAttackPiece`를 구현체를 생성하였습니다.
+ `Piece`는 `PieceType`과 `Color`를 가지고, `PieceType`은 각 타입에 따른 이동 방향과 스코어 규칙을 가집니다.

**`DirectionAttackPiece` (StraightCaptureObstacleRule에서 이름 변경) 는 그대로 유지하되, 기물의 구현체들이 `DirectionAttackPiece` 를 상속받도록 하였습니다**

### 관련 질문

제시해주신 ROOK 의 전략이 변경되는 상황에서 제가 다른 개발자의 코드를 봤었더라면,

> ROOK 클래스 확인 → DirectionAttackPiece 확인 → ROOK에 대한 전략 생성 및 상위 클래스 변경

으로 진행을 할 것 같다는 생각으로 구현을 했는데요

사실 기존 전략에 대한 부분은 최대한 가져가면서 변경을 시도하려고 해서 확신이 잘 서지 않습니다 🥲
현구막이 보시기에는 상속으로 변경했다 하더라도 이전의 설계와 크게 달라보이시지 않는지… 궁금합니다..! 

---
### 턴 도입

저번 리뷰 때 턴이 없어도 괜찮을 것 같다고 생각한다고 말씀 드렸는데요..!

이번 단계를 진행하면서, 체스는 도메인 규칙이 턴 패스를 허용하지 않는다는 것을 알게 되었습니다..!

따라서 턴을 도입하게 되었습니다… 도메인에 대한 이해가 부족했던 것 같아 아쉽습니다 😭

---

이번 단계도 잘 부탁드립니다!🙏😁

jminkkk committed:
fix: 픽스처 사용하여 변경한 지역 변수에 대해 레거시 제거

jminkkk committed:
refactor: obstacle 패키지를 domain아래로 위치 수정

jminkkk committed:
refactor: 타입 별 장애물에 대한 규칙 구현체 생성

jminkkk committed:
docs(README.md): 추가된 요구사항 명세 분석 작성 및 설계 일부분 변경

jminkkk committed:
feat(Turn): 게임 시작 시 턴은 검정팀부터 시작하는 기능 구현

jminkkk committed:
test(Turn): 해당 색상의 팀 차례인지 확인하는 테스트 추가

jminkkk committed:
refactor(Turn): 턴 전략을 받아 객체를 생성하는 로직 추가 및 검정팀 먼저인 턴 전략 구현

jminkkk committed:
refactor(Turn): 턴에 color를 받는 생성자 추가 및 턴 생성 전략 제거

jminkkk committed:
test(Turn): 다음 차례를 반환하는 기능 구현

jminkkk committed:
refactor(Color): 같은 색상인지 확인하는 메서드의 인자를 Piece에서 Color로 수정

jminkkk committed:
refactor(Turn): 누구의 차례인지 확인하는 로직에서 piece를 받도록 수정

jminkkk committed:
fix(BoardTest): 이동을 검사하는 테스트에서 턴을 고려하여 테스트 Piece 변경

jminkkk committed:
refactor(Board): 턴을 체크하는 기능 구현

jminkkk committed:
refactor(EnemyExistPolicy): 타겟이 적인지 판단하는 로직을 Piece에서 EnemyExistPolicy 내부로 이동

- Policy의 isSatisfied메서드 인자를 적 유무 여부에서 타겟 Piece로 수정

jminkkk committed:
docs(README.md): 요구사항 객체 재 설계

jminkkk committed:
refactor(PieceType): PieceType에서 ObstacleRule 제거

jminkkk committed:
refactor(obstaclerule/):  각 피스별 ObstacleRule 구현체 제거

jminkkk committed:
refactor(Position): 두 포지션의 file을 비교해 rank만 이동한 것인지 체크하는 기능을 Position으로 위치 변경

jminkkk committed:
refactor(Piece): 추상 클래스로 변경

jminkkk committed:
refactor: ObstacleRule을 Attacker로 이름 변경 및 인터페이스로 변경

jminkkk committed:
refactor(Empty): Piece를 상속 받는 Empty 객체 생성

jminkkk committed:
refactor(Empty): Empty를 싱글톤으로 사용하도록 변경

jminkkk committed:
refactor(Attacker): Attacker의 구현체들을 Attacker와 Piece을 상속받도록 변경

jminkkk committed:
refactor(Attacker): 자기자신을 제거하는 로직 Attacker에서 구현체들로 변경

jminkkk committed:
refactor(Bishop): DirectionAttackPiece를 상속받는 Bishop 객체 구현 및 관련 테스트 Piece에서 이동

jminkkk committed:
refactor(King): DirectionAttackPiece를 상속받는 King 객체 구현 및 관련 테스트 Piece에서 이동

jminkkk committed:
refactor(Knight): DirectionAttackPiece를 상속받는 Knight 객체 구현 및 관련 테스트 Piece에서 이동

jminkkk committed:
refactor(Queen): DirectionAttackPiece를 상속받는 Queen 객체 구현 및 관련 테스트 Piece에서 이동

jminkkk committed:
refactor(Rook): DirectionAttackPiece를 상속받는 Rook 객체 구현 및 관련 테스트 Piece에서 이동

jminkkk committed:
refactor(Pawn): DiagonalAttackPiece를 상속받는 Pawn 객체 구현 및 관련 테스트 Piece에서 이동

jminkkk committed:
fix(Piece): Piece가 아닌 각 말의 구현체들로 변경

jminkkk committed:
feat(Piece): 이동가능한지 확인하는 메서드를 추상 메서드로 변경 및 equals와 hashcode 구현

jminkkk committed:
refactor(piece/attacker/): AttackPiece의 구현체에 대한 패키지 위치 수정 및 이름 변경

jminkkk committed:
feat(ScoreRule): 점수 규칙에 대한 추상 클래스 생성

jminkkk committed:
feat(NormalScoreRule): 기본 점수에 곱으로 계산하는 점수 규칙 로직 구현

jminkkk committed:
feat(PawnScoreRule): 폰에 대한 점수 규칙 생성 및 같은 줄에 같은 색의 폰이 있는 경우 점수 계산 기능 구현

jminkkk committed:
test(PawnScoreRule): 같은 줄에 폰이 없는 경우 점수 계산 기능 구현

jminkkk committed:
feat(NoScoreRule): 점수가 없는 경우 0을 반환하는 NoScoreRule 구현

jminkkk committed:
feat(PieceType): 타입별 ScoreRule 구현 및 Pawn에 대한 점수 테스트 작성

jminkkk committed:
test(PieceType): 폰을 제외한 말들의 점수 테스트 작성

jminkkk committed:
feat(Referee): 팀별 피스 목록을 점수 규칙에 맞게 점수를 계산하는 Referee 객체 구현

jminkkk committed:
fix(Board): turn을 불변에서 변경

jminkkk committed:
feat(GameCommand): 현재 스코어를 확인할 수 있는 STATUS 명령어 추가

jminkkk committed:
fix(PawnScoreRule): 같은 file에 있는 개수를 계산하는 메서드 올바르게 수정

jminkkk committed:
feat(ChessGameController): 기물별 점수를 출력하는 기능 구현

jminkkk committed:
feat(OutputView): 현재 팀의 점수를 출력하는 메서드 구현

jminkkk committed:
feat(Board): King이 잡힐 시에 게임이 종료되는 기능 구현

jminkkk committed:
feat(ChessGameController): 게임이 끝났는지 확인하고 진행하도록 구현

jminkkk committed:
chore(build.gradle): mysql 의존성 추가

jminkkk committed:
feat(OuputView): king이 잡혀 게임 이 종료되었을 때 출력문 추가

jminkkk committed:
chore(docker/): docker-compose.yml 작성

jminkkk committed:
test(DatabaseConfig): 드라이버 연결 설정 파일 작성

jminkkk committed:
feat(PieceDao): Piece와 해당 Piece의 Position를 위한 Dao 구현

jminkkk committed:
feat(PieceDto): piece model, dao간 변환을 위한 dto 생성 및 File, Rank에 변환을 위한 value 추가

jminkkk committed:
feat(TurnDao): Turn를 위한 Dao 구현

jminkkk committed:
feat(TurnDto): turn model, dao간 변환을 위한 dto 생성

jminkkk committed:
test(FakeDao): 테스트용 FakeDao 생성

jminkkk committed:
feat(ddl.sql): ddl 문 추가

jminkkk committed:
feat(Board, BoardInitializer): 저장된 게임을 확인하고 여부를 받는 게임 로직 추가 및 BoardInitializer를 인터페이스로 변경

jminkkk committed:
feat(ChessGameService): Dao를 통해 DB 관련 로직을 수행하는 Service 구현

jminkkk committed:
feat(ChessGameController): 재시작 여부를 받고 저장된 게임을 실행하는 컨트롤러 로직 작성

jminkkk committed:
refactor: getter 대신 객체에게 물어보도록 수정

jminkkk committed:
fix(FakePieceDao): LinkedList로 변경

jminkkk committed:
refactor(ChessGameService): 전체 내역 삭제하는 메서드 이름 변경 및 PieceDao 필드를 인터페이스로 변경

jminkkk committed:
test(ChessGameService): Fake 객체를 활용하여 Servive 단 로직 테스트

jminkkk committed:
fix(ChessGameServiceTest): 저장된 PieceDto의 순서 상관없이 원소만 확인

jminkkk committed:
fix(ddl.sql): 불필요한 쿼리문 제거

Hyeon9mak commented on src/main/java/chess/controller/ChessGameController.java:
```diff
@@ -1,26 +1,35 @@
 import chess.view.OutputView;
+import java.util.Objects;
 
 public class ChessGameController {
 
     private final OutputView outputView;
     private final InputView inputView;
+    private final ChessGameService gameService;
```
service 라는 명칭을 사용한 이유가 있나요? 🤔

jminkkk replied:
스프링을 사용할 때, 컨트롤러와 레포지토리 사이에서 비즈니스 로직을 처리하는 Service 계층을 사용했었는데요..!

제가 생각하기에 Service 계층의 역할은 크게 2가지라고 생각합니다!
1. 비즈니스 로직 처리 
2. 로직에 맞는 DAO 호출

이번 미션에서 비즈니스 로직은 도메인에서 거의 처리가 되었기 때문에, 1번 역할을 수행하지는 않았는데요.
하지만 이번에 적용한 Service 객체가 2번 역할을 수행하는 용도와 닮아있다고 생각해서 Service라는 네이밍을 사용했습니다!

Hyeon9mak replied:
비즈니스 로직은 각각의 비즈니스 객체(domain 계층)가 수행할 수 있는 걸로 보여요.
그래야 비즈니스 변경 요구사항이 전달되었을 때 controller, service 패키지를 열어볼 필요 없이
domain 패키지 탐색을 시작으로 빠르게 수정할 대상을 찾아낼 수 있기 때문이에요.
(controller 가 view 와 domain 의 소통에만 집중해야하는 것과 같은 이유)

결국 service 계층은 2번처럼 비즈니스 로직과 영속 계층사이 소통만 담당하는 것이 이상적이겠네요!

jminkkk replied:
넵넵 그동안 service 계층에 대한 오해가 조금 있었던 것 같네요 🤣
비즈니스 로직은 각각의 비즈니스 객체가 수행한다! 명심하겠습니다 ㅎㅎ

감사합니다 현구막!

Hyeon9mak commented on src/main/java/chess/controller/ChessGameController.java:
```diff
@@ -31,22 +40,79 @@ private void checkStart(final GameCommand gameCommand) {
+        }
+    }
+
+    private void processSavedGame() {
+        String saveGameStart = inputView.getSaveGameStart();
+        if (Objects.equals(saveGameStart, "y")) {
+            Board savedBoard = new Board(new SavedBoardInitializer(gameService.findAllPiece()), gameService.findTurn());
+            outputView.printBoard(savedBoard);
+            processGame(savedBoard);
+        }
+
+        if (Objects.equals(saveGameStart, "n")) {
+            gameService.delete();
+            play();
+        }
```
동료 개발자에게 SavedGame 명령어를 변경해달라는 요구사항이 전달되면,
controller 대신 다른 클래스를 열어보지 않을까요?

사용자로부터 전달되는 명령이 "y" 인지 "n" 인지는 controller 가 몰라야하는 내용 같아요!

jminkkk replied:
앗 제가 놓쳤네요 😭 
수정하겠습니다!!

Hyeon9mak commented on src/main/java/chess/controller/ChessGameController.java:
```diff
@@ -31,22 +40,79 @@ private void checkStart(final GameCommand gameCommand) {
 
-        while (gameCommand == GameCommand.MOVE) {
-            playTurn(board);
+        while (isGameContinuing(gameCommand)) {
+            processCommand(gameCommand, board);
+            if (board.isFinish()) {
+                outputView.printFinish();
+                gameService.delete();
+                return;
+            }
+
             gameCommand = inputView.getGameCommand();
         }
 
+        restartGameIfRequested(gameCommand);
+        saveGameIfEnd(gameCommand, board);
+    }
+
+    private boolean isGameContinuing(GameCommand gameCommand) {
+        return gameCommand == GameCommand.MOVE || gameCommand == GameCommand.STATUS;
+    }
```
만약 MOVE 나 STATUS 가 변경/사라지면 controller 쪽 코드도 함께 영향을 받겠는데요,
Command 에게 질문을 던져보는 건 어떨까요?

> 객체에 메세지를 보내라.

jminkkk replied:
Command도 객체인데, 컨트롤러 계층와 뷰 계층에서는 유독 객체지향을 방심(?)하게 되는 것 같아요 🥲
수정하겠습니다 피드백 감사합니다!! 🙏

Hyeon9mak commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -1,36 +1,63 @@
 
 public class Board {
 
-    private final BoardInitializer boardInitializer;
     private final Map<Position, Piece> board;
+    private Turn turn;
 
     public Board(final BoardInitializer boardInitializer) {
-        this.boardInitializer = boardInitializer;
         this.board = boardInitializer.initialize();
+        this.turn = new Turn();
+    }
+
+    public Board(final BoardInitializer boardInitializer, final Turn turn) {
+        this.board = boardInitializer.initialize();
+        this.turn = turn;
     }
```
턴이라는 건 어찌보면 게임의 진행상태 중 일부겠는데요,
현재 체스 게임은 `시작전`, `백의 차례`, `흑의 차례`, `종료` 와 같은 상태로 구분해서 게임을 바라볼 수 있겠어요.
Board 에 산재되어 있는 상태 관련 로직들을 정리해볼 수 있는 필살기... 상태패턴이 떠오르는데요? 😄

jminkkk replied:
오! 상태패턴을 잘 몰라서 사용해볼 생각을 못했었는데, 이번 기회에 해보면 좋을 것 같아요!
바로 적용해보겠습니다!! 🫡

Hyeon9mak commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -1,36 +1,63 @@
         throw new IllegalArgumentException("이동이 불가능한 위치입니다.");
     }
 
+    private void checkTurn(final Piece piece) {
+        if (!turn.isTurn(piece)) {
+            throw new IllegalArgumentException("해당 진영의 차례가 아닙니다.");
+        }
+    }
```
부정형 질의문을 던지면 if 키워드 사이 눈에 잘 띄지 않는 `!`를 제거해볼 수 있겠어요.

jminkkk replied:
앗 제가 놓쳤네요 🥲
TDD로 미션을 진행하다보면, 메서드가 특정 조건을 충족하는지를 검증하게 돼서, 긍정형으로 메서드를 작성하게 되는 것 같아요 🥲
로직 작성할 때 더 꼼꼼히 확인하겠습니다!

Hyeon9mak commented on src/main/java/chess/domain/board/InitialPiecePosition.java:
```diff
@@ -1,41 +1,47 @@
-    BLACK_KING(new Piece(PieceType.KING, Color.BLACK), List.of(Position.of(5, 8))),
-    WHITE_KING(new Piece(PieceType.KING, Color.WHITE), List.of(Position.of(5, 1))),
+    BLACK_KING(new King(Color.BLACK), List.of(Position.of(5, 8))),
+    WHITE_KING(new King(Color.WHITE), List.of(Position.of(5, 1))),
 
-    BLACK_QUEEN(new Piece(PieceType.QUEEN, Color.BLACK), List.of(Position.of(4, 8))),
-    WHITE_QUEEN(new Piece(PieceType.QUEEN, Color.WHITE), List.of(Position.of(4, 1))),
+    BLACK_QUEEN(new Queen(Color.BLACK), List.of(Position.of(4, 8))),
+    WHITE_QUEEN(new Queen(Color.WHITE), List.of(Position.of(4, 1))),
```
훨씬 직관적인 느낌이네요 😄 👍

Hyeon9mak commented on src/main/java/chess/domain/piece/attacker/DiagonalAttackPiece.java:
```diff
@@ -1,29 +1,45 @@
 import java.util.Map;
 
-public class DiagonalCaptureObstacleRule extends ObstacleRule {
+public class DiagonalAttackPiece extends Piece implements Attacker {
+    protected DiagonalAttackPiece(final PieceType pieceType, final Color color) {
+        super(pieceType, color);
+    }
+
```
Bishop, Queen, Knight, King 같은 경우도 대각 공격이 가능하지 않나 싶은 생각이들어요.
`DirectionAttackPiece` 와 `DiagonalAttackPiece` 를 나눈 의도를 아직 잘 이해를 못했는데, 혹시 설명이 가능할까요?

jminkkk replied:
`DirectionAttackPiece` -> 이동방향과 동일하게 공격하는 피스
`DiagonalAttackPiece ` -> 이동방향과 달리, 대각 공격하는 피스

위를 기준으로 나누었는데, 네이밍에 오류가 있었던 것 같아요 🥲
수정하겠습니다!

Hyeon9mak commented on src/main/java/chess/domain/position/Position.java:
```diff
@@ -60,4 +60,9 @@ public boolean isDiagonalBy(final Position target) {
                 && rank.canMoveOneSpace(target.rank);
     }
+
+    // TODO: Test 추가
+    public boolean isRankMove(final Position target) {
+        return this.file == target.file;
+    }
```
👀

jminkkk replied:
헉 인텔리제이가 일을 안했네요 🥲
수정하겠습니다!

Hyeon9mak commented on src/test/java/chess/domain/piece/RookTest.java:
```diff
@@ -0,0 +1,90 @@
+
+class RookTest {
+
+    /*
+     * ...E....
+     * ...E....
+     * ...E.....
+     * ...E....
+     * EEErEEEE
+     * ...E....
+     * ...E....
+     * ...E....
+     * */
```
이렇게 남긴 소소한 주석들이 동료 개발자들 모두의 생산성을 크게 향상시키죠 💯

Hyeon9mak reviewed (changes requested):
안녕하세요 몰리~ 마지막 단계까지 구현하시느라 고생하셨어요.
우선 3단계 리뷰를 집중적으로 진행하고, 4단계 리뷰를 추가로 진행할게요.

### 상속으로 변경했다 하더라도 이전의 설계와 크게 달라보이시지 않는지…

제가 아직 몰리 코드에 대한 이해도 낮은 것 같아요. 🙏
`DiagonalAttackPiece` 과 `DirectionAttackPiece` 의 차이가 직접적으로 와닿지 않아서
두 구현체의 차이에 대해 질문드리는 코멘트를 남겼어요. 
답변 주시면 추가로 고민해볼게요.

### 턴 도입

체스 경기 결승전을 보면 서로 버저를 번갈아 누르며 한칸씩 슥슥삭삭 무조건 이동을 하죠 😄 👍 

흑과 백의 턴 말고도 게임 진행 전, 게임 종료 후 등의 상태도 관리가 필요해보이는데요,
관련해서 코멘트를 하나 남겼어요 😄 

우선 간단한 코멘트들 남겼으니 확인해주세요!

jminkkk committed:
test(Position): rank 이동에 대해 참거짓을 반환하는 테스트 작성

jminkkk committed:
refactor(GameCommand): GameCommand에게 게임을 진행하는 명령어인지 물어보도록 수정

jminkkk committed:
refactor(InputView): 저장된 게임을 실행하는 로직에서 입력된 문구가 일치하는지 처리를 InputView로 이동, 컨트롤러는 물어보도록 수정

jminkkk committed:
refactor(Turn): 게임의 차례를 나타내는 Turn 객체 제거

jminkkk committed:
refactor(GameStatus): 게임의 상태를 나타내기 위한 인터페이스 생성

jminkkk committed:
feat(Running): 게임이 진행 중인 상태를 나타내는 Running 추상 클래스 생성

jminkkk committed:
feat(Running): 게임 종료를 확인하는 메서드 구현

jminkkk committed:
feat(BlackTurn): 검정 팀이 차례인 상태에 따른 행위 구현

jminkkk committed:
feat(WhiteTurn): 화이트 팀이 차례인 상태에 따른 행위 구현

jminkkk committed:
feat(Finished): 종료 상태에 따른 행위 구현

jminkkk committed:
test(WhiteTurn, BlackTurn): 현재 차례에 해당하는 색을 올바르게 반환하는지 테스트

jminkkk committed:
feat(Game): 게임의 진행을 관리하는 Game 클래스 구현

jminkkk committed:
refactor(Board): 턴 및 게임의 상태를 관리하는 로직 제거

jminkkk committed:
refactor(TurnDto): Turn 객체가 아닌 색상을 받아 전달하도록 수정

jminkkk committed:
feat(OutputView, ColorDisplay): 저장된 게임 실행 시 기존 차례에 해당하는 색을 출력하는 메서드 및 ColorDisplay 이넘 추가

jminkkk committed:
refactor(ChessGameController, ChessGameService): Board가 아닌 Game을 실행하도록 변경

jminkkk committed:
fix(PieceDto): 타입에 맞는 객체 반환 시, 다른 타입으로 생성되는 피스 수정

jminkkk committed:
fix(Empty): 생성자를 private로 변경

jminkkk committed:
fix(Finished): Finished는 Running을 상속받지 않도록 수정

jminkkk committed:
fix(InputView, OutputView): 저장된 게임의 실행 여부를 묻는 출력문을 입력받기 전에 출력하도록 위치 수정

jminkkk committed:
refactor(ChessGameService): 게임을 저장하는 메서드의 이름을 변경

jminkkk committed:
fix(BoardTest): 상태패턴 도입으로 boolean을 반환하는 메서드의 테스트도 수정

jminkkk committed:
refactor(ChessGameController): 흐름이 보이도록 수정 및 추출하지 않아도 되는 메서드 합치기

jminkkk committed:
refactor(ObstacleFinder): Attacker 에서 ObstacleFinder로 이름 변경

jminkkk committed:
refactor(AttackPiece): AttackPiece 구현체 제거 및 ObstacleRule만 상속받도록 수정

jminkkk committed:
refactor(MovementDirectionObstacleFinder): MovementDirectionObstacleFinder를 사용하는 비숍, 킹, 퀸, 룩 구현

jminkkk committed:
refactor(Empty): Piece를 상속받도록 수정

jminkkk committed:
refactor(Pawn): PawnObstacleFinder를 사용하여 움직임을 판단하도록 수정

jminkkk committed:
fix(KnightTest): 시작 위치가 보드 목록에서 누락되어 있던 것 수정

jminkkk committed:
fix(KnightObstacleFinder, Knight): 타겟을 비교하여 나이트의 규칙에 맞는 장애물을 반환하는 KnightObstacleFinder 구현

jminkkk committed:
fix(Knight): 잘못 임포트된 패키지 변경

jminkkk committed:
refactor(Piece): 같은 팀인지 확인하는 메서드 긍정형으로 변경

jminkkk committed:
fix(game/): 게임 패키지 생성 및 board 패키지에서 game 관련 객체 이동

- obstaclefinder 패키지 명을 변경된 인터페이스 명에 맞게 변경
- 깃 충돌로 남아있던 attacker 삭제 및 내부 구현체 이동

jminkkk committed:
fix(ChessGameController): start 여러번 수행 시 end가 되지 않는 이슈 해결 및 전체 메서드명 일관성있게 수정

jminkkk reviewed (commented):
안녕하세요 현구막 🙌
3단계 리뷰 재요청 드립니다 ㅎㅎ

### 상태 패턴

지난 리뷰 때 현구막께서 상태패턴을 추천해주셔서 반영을 해보았습니다..!

<img src = https://github.com/woowacourse/java-chess/assets/102847513/44e4d69c-bcf7-4f94-9169-20b3c7f7456b width = 50%>

상태 패턴을 도입하다보니 자연스래 게임 진행과 보드 관리에 대한 부분이 분리가 되어서 상태패턴의 장점을 느낄 수 있었던 것 같아요 😆

### Attacker 관련

기존에 Attacker 인터페이스를 동일한 공격 정책(?)을 가진 기물들이 상속을 받도록 구현을 했는데요..!
현구막에게 설계에 대해서 질문을 드리고 다시 돌아보니… 제가 봐도 이해하기 어려운 코드 같더라구요…! 🥲

그래서 아래처럼 변경을 시도해보았습니다..!

---

> Attacker라는 이름을 가진 인터페이스가 findObstacle이라는 책임을 수행하는 게 낯설어보일 수 있겠다는 생각이 들었어요. 
또, 구현체들(`DirectionAttackPiece` ,`DiagonalAttackPiece`)의 이름 또한 의미전달이 잘되지 않는 것 같다고 느꼈어요...!

1. Attacker를 ObstacleFinder로 이름을 변경하고,
2. `extends Piece implements Attacker`를 구현했던 객체들 대신, 다음처럼 사용하도록 했어요
  +  Pawn, Knight처럼 장애물 판단 로직이 다른 객체들에 대한 PawnObstacleFinder, KnightObstacleFinder를 구현
 + 이동 방향으로 장애물을 찾는 나머지 객체들은 MovementDirectionObstacleFinder라는 객체를 사용하도록 함

----

> 기존 Piece의 이동 판단 로직에 대해서 생각을 해보았어요. 상위 클래스에서 전부 처리하는 것보다는 각 기물들이 자신이 가지고 있는 ObstacleFinder를 이용하여 장애물 목록을 받게하면 책임을 나눌 수 있지 않을까 하는 생각을 했어요..!

+ Composition을 사용해 수정하여, 각 기물들이 생성 시에 자신에게 맞는 ObstacleFinder를 생성하도록 했어요.
+ 이동 판단 시에는 장애물 목록을 각자의 규칙에 맞게 생성하여 반환하도록 하였습니다...!

---

설계쪽 수정이 잦아서 이해하기 힘드실 수 있을 것 같아요 😭
이번 리뷰도 잘 부탁드립니다... 🫡

jminkkk committed:
refactor(ChessGameController): 게임이 진행중인지 확인하는 메서드명 수정

Hyeon9mak commented on src/main/java/chess/controller/ChessGameController.java:
```diff
@@ -31,31 +44,84 @@ private void checkStart(final GameCommand gameCommand) {
+    }
+
+    private boolean isGameActionIfOngoing(final Game game, final GameCommand gameCommand) {
+        return !game.isFinish()
+                && gameCommand != GameCommand.START
+                && gameCommand != GameCommand.END;
+    }
+
+    private void executeMoveCommand(final Game game, final GameCommand gameCommand) {
+        if (gameCommand == GameCommand.MOVE) {
+            executeTurn(game);
+        }
+    }
+
+    private void executeStatusCommand(final Game game, final GameCommand gameCommand) {
+        if (gameCommand == GameCommand.STATUS) {
+            viewScore(game.getBoard());
+        }
+    }
+
+    private void checkGameRestart(final GameCommand gameCommand) {
         if (gameCommand == GameCommand.START) {
-            play();
+            playNewGame();
+        }
```
역시나 메세지를 보내서 해결할 수 있겠어요.

Hyeon9mak commented on src/main/java/chess/domain/game/status/GameStatus.java:
```diff
@@ -0,0 +1,16 @@
+import chess.domain.piece.Color;
+import chess.domain.position.Position;
+
+public interface GameStatus {
+
+    GameStatus move(final Position source, final Position target);
+
+    boolean isFinish();
+
+    Color getTurn();
+
+    Board getBoard();
+}
```
👍 디자인 패턴을 꼭 알 필요는 없지만 알아둔다면 동료 개발자들과의 소통이 훨씬 수월할거에요.

> 물론 그렇다고 아키텍쳐와 디자인 패턴에 대한 공부를 소홀히 해도 된다는 것은 아니다. 특히나 개발자는 이직과 팀 변경이 잦기 때문에 팀 컨벤션을 길게 유지하기가 쉽지 않다. 대중적으로 갖춘 배경 지식을 기반으로 빠르게 코드를 파악하기 위해서, 빠르게 파악할 수 있는 코드를 작성해주기 위해서. 서로 일을 잘하기 위한 상호협약, 예의범절 같은 것이라고 생각한다.

https://hyeon9mak.github.io/consider-oop/#-%EA%B0%9D%EC%B2%B4%EC%A7%80%ED%96%A5%EC%9D%80-%EC%98%88%EC%88%A0%EC%9D%B4-%EC%95%84%EB%8B%88%EC%95%BC 😄

jminkkk replied:
> 서로 일을 잘하기 위한 상호협약, 예의범절 같은 것이라고 생각한다.

위 문장이 인상깊네요! '객체지향은 예술이 아니야' 부분두요😆
좋은 글 감사합니다 현구막!🫡

Hyeon9mak commented on src/main/java/chess/domain/game/Game.java:
```diff
@@ -0,0 +1,51 @@
+        this.gameStatus = gameStatus;
+    }
+
+    public static Game of(final BoardInitializer boardInitializer, final Color turnColor) {
+        return new Game(findGameStatus(boardInitializer, turnColor));
+    }
+
+    private static GameStatus findGameStatus(final BoardInitializer boardInitializer, final Color turnColor) {
+        if (turnColor.isSameColor(Color.BLACK)) {
+            return new BlackTurn(new Board(boardInitializer));
+        }
+
+        return new WhiteTurn(new Board(boardInitializer));
+    }
+
+    public void move(final Position source, final Position target) {
+        gameStatus = gameStatus.move(source, target);
+    }
+
+    public boolean isFinish() {
+        return gameStatus.isFinish();
+    }
+
+    public Map<Position, Piece> getBoard() {
+        return gameStatus.getBoard().getBoard();
+    }
+
+    public Color getTurn() {
+        return gameStatus.getTurn();
+    }
```
덕분에 복잡한 분기로직이 많아 사라졌네요!!

jminkkk replied:
상태패턴의 장점을 확실히 경험할 수 있었어요 👍

Hyeon9mak commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,18 @@
+version: "3.9"
+services:
+  db:
+    image: mysql:8.0.28
+    platform: linux/x86_64
+    restart: always
+    ports:
+      - "13306:3306"
+    environment:
+      MYSQL_ROOT_PASSWORD: root
+      MYSQL_DATABASE: chess
+      MYSQL_USER: user
+      MYSQL_PASSWORD: password
+      TZ: Asia/Seoul
+    volumes:
+      - ./db/mysql/data:/var/lib/mysql
```
타임존 설정까지 👍

Hyeon9mak commented on src/main/java/chess/repository/DatabaseConfig.java:
```diff
@@ -0,0 +1,24 @@
+import java.sql.DriverManager;
+import java.sql.SQLException;
+
+public final class DatabaseConfig {
+
+    private static final String SERVER = "localhost:13306";
+    private static final String DATABASE = "chess";
+    private static final String OPTION = "?useSSL=false&allowPublicKeyRetrieval=true&serverTimezone=UTC";
+    private static final String USERNAME = "root";
+    private static final String PASSWORD = "root";
+
+    public static Connection getConnection() {
```
docker-compose.yml 에 기입된 설정과 많이 달라보여요. username 과 password, timezone 이 모두 다르네요 🤔

jminkkk replied:
~~앗 강의 코드를 참고했는데.. 폐해네요...😭~~

~~root로 접속하고 있어서, 지금 username 과 password은 굳이 필요하지 않을 것 같아요!~~
~~수정하겠습니다...!!!~~

---

네오가 왜 코드에 해당 부분을 작성했을지, 저 부분을 작성하게 되면 무슨일이 생기는지 생각을 해보았어요!

+ docker-compose.yml
```mysql
      MYSQL_USER: user
      MYSQL_PASSWORD: password
```
일단 docker-compose.yml에 해당 부분을 작성하게 되면 user라는 이름을 가진 유저가 데이터베이스에 생기게 된다는 사실을 알았습니다.

> 그렇다면 root로 접속하면 되지 왜 user를 생성하지?

라는 생각을 했습니다. 이 부분은 크루들에게 의견을 들었는데요.
root로 접속하게 되면 모든 데이터베이스에 대한 권한을 가지고 있기 때문에 위험할 수 있다. (실수로 삭제를 한다던가 하는 등)
때문에 해당 DB를 관리하는 유저를 하나 생성하고 그 유저만 해당 DB에 접근하여 관리하도록 하는 것이 안전하다라는 이유를 들을 수 있었습니다!

따라서 다시 user라는 유저를 통해 접근하도록 수정하였습니다!
~~무지성 참고를 방심하지 않도록 해야겠어요...🥲~~

Hyeon9mak commented on src/main/java/chess/service/ChessGameService.java:
```diff
@@ -0,0 +1,60 @@
+                .collect(Collectors.toMap(PieceDto::getPositionFrom, PieceDto::getPieceFrom));
+    }
+
+    public Color findTurn() {
+        TurnDto turnDto = turnDao.findAll().get(0);
+        return turnDto.from();
+    }
```
영속 대상으로부터 단건만 찾아내면 되지 않을까요? 🤔

Hyeon9mak commented on src/main/java/chess/service/ChessGameService.java:
```diff
@@ -0,0 +1,60 @@
+        turnDao.deleteAll();
+    }
+
+    public boolean existSavedGame() {
+        return findAllPiece().size() != 0;
+    }
```
영속 대상에게 값이 1건이라도 있냐고 물어볼 수 있겠어요!

jminkkk replied:
앗 바로 수정하겠습니다 🫡

Hyeon9mak commented on src/main/java/chess/repository/turn/FakeTurnDao.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class FakeTurnDao implements TurnDao {
+    private List<TurnDto> turnDtos = new ArrayList<>();
+
+    @Override
+    public void save(final TurnDto turnDto) {
+        turnDtos.add(turnDto);
+    }
+
+    @Override
+    public List<TurnDto> findAll() {
+        return turnDtos;
+    }
+
+    @Override
+    public void deleteAll() {
+        turnDtos = new ArrayList<>();
+    }
```
인터페이스로 분리해주신 덕분에 테스트 더블을 만들기 편리하네요! 👍 
그런데.. 테스트에만 사용되는 테스트 더블...
그렇다면 위치를 main.java 대신 다른 곳으로 옮겨보면 어떨까요?

jminkkk replied:
앗 간과했네요 ㅠㅠ 
test/ 아래로 수정하겠습니다!!

Hyeon9mak commented on src/main/java/chess/repository/turn/TurnDao.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.util.List;
+
+public interface TurnDao {
+
+    void save(final TurnDto turnDto);
+
+    List<TurnDto> findAll();
+
+    void deleteAll();
```
인터페이스가 DAO(DataAccessObject) 라는 명칭을 꼭 알 필요가 있을까요? 🤔 
구현체가 어떤 형상을 가졌는지 노출되는 것 같아요.

jminkkk replied:
저는 일단 TurnDao 클래스만을 사용하다가, 해당 클래스의 테스트를 위해 FakeTurnDao를 추가하려고 하면서 인터페이스를 하게 되었는데요..!

즉, DAO를 추상화 하려는 의도로 인터페이스를 생성했기 때문에 인터페이스의 이름을 DAO로 작성을 했습니다...!
그런 DAO의 구현체이니 당연히 DAO라는 것을 알아도 괜찮지 않을까 라는 생각입니다...!

그런데, 사실 제가 현구막의 질문을 잘 이해하지 못한 것 같아요...🥲
제가 답변한 내용이 현구막의 의도한 부분이 맞을까요...??

Hyeon9mak commented on src/main/java/chess/service/ChessGameService.java:
```diff
@@ -0,0 +1,60 @@
+                .collect(Collectors.toMap(PieceDto::getPositionFrom, PieceDto::getPieceFrom));
+    }
+
+    public Color findTurn() {
+        TurnDto turnDto = turnDao.findAll().get(0);
+        return turnDto.from();
```
dto 는 영속화를 위해서만 사용되는 객체인데요, dto 가 현재 service 계층까지 노출되고 있어요 😱 
추후 영속방식이 변경되면 service 계층의 코드까지 변경의 영향을 받겠네요.
service 는 순수한 domain 정보를 넘기고, persistence 계층에서 이를 변환하여 처리하는 것은 어떨까요?

jminkkk replied:
앗 제가 Dao와 Service의 책임을 제대로 이해하지 못했던 것 같아요
수정하겠습니다🫡

Hyeon9mak commented on src/main/java/chess/domain/piece/obstaclefinder/KnightObstacleFinder.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+import java.util.Map;
+
+public class KnightObstacleFinder implements ObstacleFinder {
+
+    @Override
+    public List<Position> findObstacle(final Position source, final Position target,
+                                       final Map<Position, Piece> pieces) {
+        Piece sourcePiece = pieces.get(source);
+        Piece targetPiece = pieces.getOrDefault(target, Empty.getInstance());
+
+        if (sourcePiece.isSameTeamColor(targetPiece.getColor())) {
+            return List.of(target);
+        }
+        return List.of();
+    }
+}
```
특수한 전략을 가진 pawn, knight 는 이해를 돕기 위해 별도로 구현체를 분리하고,
나머지 통상적인 전략을 가진 기물들은 통합을 해주셨군요 👍 
훨씬 이해하기 좋은 코드가 되었네요!!

Hyeon9mak commented on src/main/java/chess/repository/piece/PieceDaoImpl.java:
```diff
@@ -0,0 +1,63 @@
+    @Override
+    public void save(final PieceDto pieceDto) {
+        final var query = "INSERT INTO piece (color, pieceType, `rank`, file) VALUES(?, ?, ?, ?)\n";
+        try (final var connection = getConnection();
+             final var preparedStatement = connection.prepareStatement(query)) {
+            preparedStatement.setString(1, pieceDto.color());
+            preparedStatement.setString(2, pieceDto.pieceType());
+            preparedStatement.setInt(3, pieceDto.rank());
+            preparedStatement.setInt(4, pieceDto.file());
+            preparedStatement.executeUpdate();
+        } catch (final SQLException e) {
+            throw new RuntimeException("[METHOD] save [TABLE] piece", e);
+        }
```
try with resource 활용 👍

Hyeon9mak replied:
https://hyeon9mak.github.io/easy-insert-with-simplejdbcinsert/
소소하지만 이런 것도 재밌습니다 😄 👍

jminkkk replied:
오 KeyHolder 라는 개념을 처음 알게 되었네요! 신기해요 😮

사실 이번 미션에서 스키마를 구상하면서, TurnDao에서 단건 조회가 아닌 전체 조회로 처음 구현한 이유가 Id 값이AUTO_INCREMENT이기 때문이였어요.
AUTO_INCREMENT한 Id값을 어떤 식으로 내가 알 수 있는거지?라는 생각을 했는데, 당장은 저장 시 항상 모든 데이터를 지우고 저장하는 로직이라, 항상 한개만 있을거라 생각했고 의문이 들긴 했지만 제출이 급해 그냥 넘어갔습니다 😂

현구막의 글을 통해 의문을 해결할 수 있었네요!! 감사합니다 ㅎㅎ
Spring을 사용할 때 꼭 활용해보겠습니다 🙇

Hyeon9mak commented on src/main/java/chess/repository/piece/PieceDaoImpl.java:
```diff
@@ -0,0 +1,63 @@
+
+            while (resultSet.next()) {
+                mapToPieces(pieceDtos, resultSet);
+            }
+        } catch (final SQLException e) {
+            throw new RuntimeException("[METHOD] findAll [TABLE] piece", e);
+        }
```
예외를 잡아서 다시 다른 예외로 변환시킬 필요가 있을까요? 🤔 
오히려 SQLException 이 발생했다는 걸 인지하기가 더 어려울 수도 있겠어요.
몰리의 생각은 어떠신가요?

jminkkk replied:
CheckedException이 발생했을 때 2가지를 할 수 있다고 생각해요.

> 1. 다른 곳으로 throw 전파
> 2. try-catch 

1번처럼 SQLException(CheckedException)을 다른 메서드로 던져야하는데, 그렇게 되면 진짜 예외가 발생한 코드가 어디인지 확인하기 어렵다고 생각했습니다!

때문에 2번처럼 try-catch로 감쌌어요.
여기서도 2가지 방법을 생각했어요!

> 1. Exception 이유를 출력하고 아예 터트리지 않기
> 2. 복구를 통한 처리
> 3. RuntimeException으로 예외 변환

1번처럼 하게 될 경우, 프로그램 자체가 정상적이지 못하다고 생각했어요. 예외가 발생했음에도 프로그램이 계속 실행되기 때문에 안전하지 못하다는 문제점이 있다고 판단했습니다!

2번은 SQLException은 애플리케이션 단에서 복구하기 힘든 예외라고 생각했습니다.

때문에 3번을 선택하여 '처리가 필수이지 않은 RuntimeException(UncheckedException)을 사용하여 예외가 발생했음을 알리고 해당 예외가 SQLException인지를 출력한다면 괜찮지 않을까' 라고 생각했습니다!

Hyeon9mak reviewed (changes requested):
안녕하세요 몰리! 피드백 잘 반영해주셨어요.
설계에 대한 고민이 많기 때문에 변화가 많았던 것 같은데, 
그 덕분에 훨씬 이해하기 쉽고 유지보수 하기 좋은 형태가 된 것 같아요 😄 👍 

이번 리뷰에는 4단계도 포함시켰는데요,
테이블 설계는 건너뛰고 데이터 저장 방식 관련된 내용만 코멘트를 남겼어요.
다음 리뷰에 테이블 설계도 이야기 나눠보시죠 😄

jminkkk committed:
refactor(GameCommand): GameCommand에게 특정 명령어인지 질문하도록 수정

jminkkk committed:
refactor(FakeDao): 테스트를 위한 Fake 객체의 위치를 test/아래로 이동

jminkkk committed:
refactor(DatabaseConfig): db 타임존 Asia/Seoul로 설정

jminkkk committed:
refactor(docker-compose.yml): 불필요한 db 유저 정보 제거

jminkkk committed:
feat(PieceDao): Piece 테이블에 로우가 하나라도 존재하는지 조회하는 기능 추가

jminkkk committed:
test(PieceDao): Pieces가 DB에 존재하는지 묻는 테스트 작성 및 FakePieceDao에 existPieces 구현

jminkkk committed:
feat(ChessGameService): 저장된 게임이 존재하는지 확인하는 메서드를 existPieces를 사용해 변경

jminkkk committed:
refactor(TurnDao): 턴 조회 시 단건 조회로 변경 및 TurnDao에서 턴 조회 시 Dto가 아닌 Color로 반환하도록 수정

- TurnDto 제거

jminkkk committed:
refactor(JdbcPieceDao, JdbcTurnDao):Dao의 구현체의 이름을 의미를 명확하게 이름 변경

jminkkk committed:
refactor(PieceDto): PieceDto 제거 및 기물 타입에 맞는 피스를 생성해주는 역할을 PieceFactory에게 변경

- PieceDao에서 피스 조회 시 Map<Position, Piece>를 반환하도록 변경

jminkkk committed:
fix(JdbcPieceDao): 피스가 존재하는지 확인하는 Dao 메서드를 resultSet.next() 사용 후 값을 가져오도록 수정

jminkkk committed:
refactor(ddl.sql): 도커 실행 시 init되도록 ./db/mysql/init 아래로 이동 및 user에게 권한 부여 쿼리문 추가

jminkkk committed:
refactor(DatabaseConfig): root가 아닌 user라는 유저를 통해 접근하도록 다시 수정

- docker-compose.yml에도 관련 유저정보 추가

jminkkk committed:
style(DaoTest): 테스트 설명을 위한 DisplayName 추가

jminkkk reviewed (commented):
안녕하세요 현구막!

docker나 jdbc를 많이 사용해보지 않아서 사용법을 정확히 모르고 사용했던 부분들이 있었던 것 같아요😅
이번 리뷰도 잘 부탁드립니다 ㅎㅎ 🙇

jminkkk committed:
refactor(JdbcDao): SQLException 발생 시 기존 예외 메시지도 함께 출력하도록 수정

Hyeon9mak commented on src/main/java/chess/repository/DatabaseConfig.java:
```diff
@@ -0,0 +1,24 @@
+import java.sql.SQLException;
+
+public final class DatabaseConfig {
+
+    private static final String SERVER = "localhost:13306";
+    private static final String DATABASE = "chess";
+    private static final String OPTION = "?useSSL=false&allowPublicKeyRetrieval=true&serverTimezone=Asia/Seoul";
+    private static final String USERNAME = "user";
+    private static final String PASSWORD = "password";
+
```
> root로 접속하게 되면 모든 데이터베이스에 대한 권한을 가지고 있기 때문에 위험할 수 있다. (실수로 삭제를 한다던가 하는 등)
때문에 해당 DB를 관리하는 유저를 하나 생성하고 그 유저만 해당 DB에 접근하여 관리하도록 하는 것이 안전하다라는 이유를 들을 수 있었습니다!

👍

Hyeon9mak commented on src/main/java/chess/repository/turn/TurnDao.java:
```diff
@@ -0,0 +1,13 @@
+import chess.domain.piece.Color;
+import java.util.Optional;
+
+public interface TurnDao {
+
+    void save(final String color);
+
+    Optional<Color> findAny();
+
+    void deleteAll();
+}
```
DAO(DataAccessObject) 는 데이터베이스 혹은 데이터 영속을 관리하는 구현체를 표현하는 객체라는 뜻인데,
FakeTurnDao 의 경우는 실제 구현 내용이 영속화와 관련되어 있지 않고, 
HashMap 을 이용해 영속행위를 묘사하고 있는 것 같아요.

때문에 DAO 대신 패키지 명처럼 저장소(Repository) 라는 명칭을 사용해보는 건 어떨까? 하는 의도의 코멘트였어요! 😄

Hyeon9mak replied:
인터페이스를 설계하는 과정에 대해서는 짚고 넘어가고 싶은데요,
인터페이스를 설계하는 관점은 구현체가 아니라 사용처가 되어야 합니다.

- DAO 구현체를 추상화 하기 위해 인터페이스를 사용한다 -> X
- service 계층에서 영속화된 정보를 가져오기 위해 인터페이스를 설계하고, 그에 맞춰 DAO 로 인터페이스 설계 사항을 구현한다. -> O

인터페이스가 구현체의 관점에서 명세되면, 구현체가 변경될 때 인터페이스도 변경되고,
그 인터페이스를 의존하는 사용처도 함께 변경되어야 하기 떄문이에요.
그렇다면 인터페이스를 사용하는 의미가 희석되겠죠..!

jminkkk replied:
> service 계층에서 영속화된 정보를 가져오기 위해 인터페이스를 설계하고, 그에 맞춰 DAO 로 인터페이스 설계 사항을 구현한다.

아하 넵넵 이해했습니다! ㅎㅎ
반영하겠습니다! 🙇

Hyeon9mak commented on src/main/java/chess/service/ChessGameService.java:
```diff
@@ -0,0 +1,53 @@
+        this.turnDao = turnDao;
+    }
+
+    public Map<Position, Piece> findAllPiece() {
+        return pieceDao.findAllPiece();
+    }
+
+    public Optional<Color> findTurn() {
+        return turnDao.findAny();
+    }
+
+    public void saveGame(final Game game) {
+        delete();
+        saveAllPiece(game.getBoard());
+        saveTurn(game.getTurn());
+    }
+
+    private void saveAllPiece(final Map<Position, Piece> pieces) {
+        pieces.entrySet()
+                .forEach(entry -> pieceDao.save(entry.getValue(), entry.getKey()));
+    }
+
+    private void saveTurn(final Color color) {
+        turnDao.save(color.name());
+    }
+
+    public void delete() {
+        pieceDao.deleteAll();
+        turnDao.deleteAll();
+    }
```
service 계층이 persistence 계층에 명령을 내리는 것만 깔끔하게 집중하네요 😄 👍

Hyeon9mak commented on src/main/java/chess/repository/piece/JdbcPieceDao.java:
```diff
@@ -0,0 +1,89 @@
+    @Override
+    public void save(final Piece piece, final Position position) {
+        final var query = "INSERT INTO piece (color, pieceType, `rank`, file) VALUES(?, ?, ?, ?)";
+        try (final var connection = getConnection();
+             final var preparedStatement = connection.prepareStatement(query)) {
+            preparedStatement.setString(1, piece.getColor().name());
+            preparedStatement.setString(2, piece.getPieceType().name());
+            preparedStatement.setInt(3, position.rank().getValue());
+            preparedStatement.setInt(4, position.file().getValue());
+            preparedStatement.executeUpdate();
+        } catch (final SQLException e) {
+            throw new RuntimeException("[METHOD] save [TABLE] piece" + e.getMessage(), e);
+        }
+    }
```
> 때문에 3번을 선택하여 '처리가 필수이지 않은 RuntimeException(UncheckedException)을 사용하여 예외가 발생했음을 알리고 해당 예외가 SQLException인지를 �출력한다면 괜찮지 않을까' 라고 생각했습니다!

👍 👍 💯 
SQL Exception 이 checked exception 을 날린다는 사실을 망각하고 잘못된 코멘트를 드렸어요 🥲 🙏 
개떡같은 코멘트를 찰떡같이 해석해주셨네요.. 👍 

영속 프레임워크 내부에서도 SQLException 을 감싸서 다시 RuntimeException 으로 던져주곤 하더라구요.
감싸주신 의도가 궁금했는데, 고민 과정까지 설명해주셔서 바로 이해되었어요 👍

jminkkk replied:
저도 한번 더 고민하고 정리해볼 수 있어서 좋았습니다 👍🙇

Hyeon9mak commented on docker/db/mysql/init/ddl.sql:
```diff
@@ -0,0 +1,21 @@
+SHOW TABLES in chess;
+USE chess;
+
+CREATE TABLE Turn
+(
+    turn_id INT AUTO_INCREMENT PRIMARY KEY,
+    turn    VARCHAR(255) NOT NULL
+);
+
+CREATE TABLE Piece
+(
+    piece_id  INT AUTO_INCREMENT PRIMARY KEY,
+    color     VARCHAR(255) NOT NULL,
+    pieceType VARCHAR(255) NOT NULL,
+    `rank`    INT          NOT NULL,
+    file      INT          NOT NULL
+);
```
현재는 게임의 턴과 기물별 위치를 저장하고 있는데요,
그 외에도 게임 명령어가 어떻게 입력되었는지를 적재하는 방식도 있을 것 같아요.

몰리가 선택해주신 방법과 위 방법은 각각 어떤 장단점이 있고, 어떤 상황에 적용이 적합할까요?
혹은 두 가지 모두 저장하는 일도 있을까요?

jminkkk replied:
제가 생각하기에는 다음과 같습니다.
**[게임의 턴과 기물별 위치만 저장하는 방식]**
+ 장점
  + 현재 상태만 저장하기 때문에 관리해야 하는 데이터의 양이 적다.
  + 바로 데이터를 불러와서 시작할 수 있어 초기 시작 시간이 짧다.
+ 단점
  + 어떤 차례를 통해 현재 상태가 되었는지 알 수 없다.
  + 요구사항이 변경되어, 무르기 기능이나 퀸의 첫 이동 체크 등이 추가되었을 경우 히스토리가 없기 때문에 구현이 곤란할 것 같다.
+ 어떤 상황에 적합할까
  + 히스토리가 필요하지 않는 요구사항일 경우
  + 게임의 진행 상태만 빠르게 복구하려는 경우

**[게임 명령어가 어떻게 입력되었는지를 적재하는 방식]**
+ 장점
  + 게임의 전체 내역을 관리할 수 있다.
  + 오히려 모든 명령어를 저장하기 때문에 턴이나, 기타 다른 정보를 저장하지 않아도 될 수 있다.
+ 단점
  + 게임 시작 시 매번 전체 명령어를 실행을 시켜야 하기 때문에 리소스 낭비나 초기 시작 시간이 발생할 수 있다.
+ 어떤 상황에 적합할까
  + 실행 순서가 필요한 요구사항일 경우(ex. 앞서 말한 무르기, 퀸의 첫 이동 체크 등등)

**[두 가지 모두 저장]**
게임을 빠르게 진행하면서 게임의 진행 과정을 상세히 기록할 수 있는 장점을 모두 활용하려고 할 경우, 두 가지 모두를 저장할 수 있을 것 같습니다..!

또 실행 순서가 필요한 요구사항으로 변경되지 않는다는 보장이 없기 때문에 가능한 둘 다 저장하는 것이 좋을 것 같다는 생각이 듭니다..!

Hyeon9mak replied:
"하나의 게임진행 데이터지만 필요에 의해 여러가지 방식으로 나누어 적재를 할 수도 있다." 라는 메세지를 전달드리고 싶었어요 👍 
추후 프로젝트를 진행할 때도 이 점을 참고해서 시스템 복잡도를 낮추는 경험을 해보시면 좋겠네요 😄 👍

Hyeon9mak reviewed (changes requested):
안녕하세요 몰리! 리뷰가 많이 늦었어요. 🙏 
피드백 역시 잘 반영해주셨습니다 💯 

데이터베이스 테이블에 관련해서는 간단하게 이야기만 나누고 넘어가도 좋을 것 같아요.
히스토리를 적재하는 방식에 대한 질문을 남겼는데, 답변해주시면 될 것 같습니다!

끝이 보이네요! 마지막까지 화이팅입니다!!

jminkkk committed:
refactor(repository/): dao 명칭을 repository로 변경

jminkkk reviewed (commented):
안녕하세요 현구막!
마지막까지 리뷰 잘 부탁드립니다 🙇 

(ps. 요즘 독감이 유행인데, 감기 조심하세요!😀)

Hyeon9mak reviewed (approved):
안녕하세요 몰리! 마지막 미션까지 정말 고생 많으셨어요.
체스 미션동안 남겨드린 코멘트들이 몰리께 도움이 되셨을까요?

요구사항도 모두 만족되었고, 답변도 잘 해주셔서 미션 마무리하려고 합니다.
낮시간 날씨도 따듯하고, 벚꽃도 피는 기분 좋은 방학이네요.
너무 공부에만 올인하기보단, 크루들과 함께 벚꽃 구경도 하시면서 추억 많이 쌓으시고
여유롭게 다음 레벨 잘 준비하시면 좋겠습니다.

고생 많으셨어요!
